app/main/helpers/validation.py

```diff
@@ -0,0 +1,185 @@
+            if validation['name'] == message_key:
+                return validation['message']
+        default_messages = {
+            'answer_required': 'Answer required'
+        }
+        return default_messages.get(
+            message_key, 'There was a problem with the answer to this question')
+
+    def errors(self):
+        raise NotImplementedError("only a subclass should be used")
+
+    def all_fields(self):
+        return reduce(add, (section.get_question_ids() for section in self.content))
+
+    def fields_with_values(self):
+        # empty list types will not be included so only string types need to be considered
+        return set(key for key, value in self.answers.items()
+                   if not isinstance(value, six.string_types) or len(value) > 0)
+
+    def character_limit_errors(self):
+        errors_map = {}
+        for question_id in self.all_fields():
+            if self.content.get_question(question_id).get('type') in ['text', 'textbox_large']:
+                answer_length = len(self.answers.get(question_id, ""))
+                if (self.character_limit is not None and answer_length > self.character_limit):
+                    errors_map[question_id] = "under_character_limit"
+
+        return errors_map
+
+    def formatting_errors(self, answers):
+        errors_map = {}
+        if self.email_validation_fields is not None and len(self.email_validation_fields) > 0:
+            for field in self.email_validation_fields:
+                if not re.match(EMAIL_REGEX, self.answers.get(field, '')):
+                    errors_map[field] = 'invalid_format'
+        return errors_map
+
+    def get_required_fields(self):
+        try:
+            req_fields = self.required_fields
+        except AttributeError:
+            req_fields = set(self.all_fields())
+
+        #  Remove optional fields
+        if self.optional_fields is not None:
+            req_fields -= set(self.optional_fields)
+
+        return req_fields
+
+
+class EmptyValidator(DeclarationValidator):
+    """
+    An empty validator that will always pass.
+    """
+    required_fields = set()
+
+    def errors(self):
+        return {}
+
+
+class G7Validator(DeclarationValidator):
+    """
+    Validator for G-Cloud 7.
+    """
+    optional_fields = set([
+        "SQ1-1p-i", "SQ1-1p-ii", "SQ1-1p-iii", "SQ1-1p-iv",
+        "SQ1-1q-i", "SQ1-1q-ii", "SQ1-1q-iii", "SQ1-1q-iv", "SQ1-1cii", "SQ1-1i-ii",
+        "SQ1-1j-i", "SQ1-1j-ii", "SQ4-1c", "SQ3-1k", "SQ1-1i-i"
+    ])
+    email_validation_fields = set(['SQ1-1o', 'SQ1-2b'])
+    character_limit = 5000
+
+    def errors(self):
+        errors_map = {}
+        errors_map.update(self.character_limit_errors())
+        errors_map.update(self.formatting_errors(self.answers))
+        errors_map.update(self.answer_required_errors())
+        return errors_map
+
+    def answer_required_errors(self):
```

Comment:
I'm proposing to move `answer_required_errors`, which wouldn't be called in the parent class since it's only called in `errors`